openssl/src/ssl/mod.rs

```diff
@@ -3503,6 +3712,190 @@ impl SslRef {
+        }
+    }
+
+    /// Adds name as an additional reference identifier that can match the peer's certificate. Any
+    /// previous names set via SSL_set1_host() or SSL_add1_host() are retained, no change is made
+    /// if name is NULL or empty. When multiple names are configured, the peer is considered
+    /// verified when any name matches. This function is required for DANE TLSA in the presence of
+    /// service name indirection via CNAME, MX or SRV records as specified in RFC7671, RFC7672 or
+    /// RFC7673.
+    ///
+    /// Requires OpenSSL 1.1.0 or newer.
+    #[corresponds(SSL_add1_host)]
+    #[cfg(ossl110)]
+    pub fn add1_host(&mut self, hostname: &str) -> Result<(), ErrorStack> {
+        let cstr = CString::new(hostname).unwrap();
+        unsafe { cvt(ffi::SSL_add1_host(self.as_ptr(), cstr.as_ptr())).map(|_| ()) }
+    }
+
+    /// Must be called before the SSL handshake is initiated with SSL_connect(3)
+    /// if (and only if) you want to enable DANE for that connection.
+    ///
+    /// The connection must be associated with a DANE-enabled SSL context.
+    ///
+    /// The base_domain argument specifies the RFC7671 TLSA base domain,
+    /// which will be the primary peer reference identifier for certificate name checks.
+    ///
+    /// Additional server names can be specified via `add1_host`.
+    ///
+    /// The base_domain is used as the default SNI hint if none has yet been
+    /// specified via SSL_set_tlsext_host_name(3).
+    ///
+    /// Requires OpenSSL 1.1.0 or newer.
+    #[corresponds(SSL_dane_enable)]
+    #[cfg(ossl110)]
+    pub fn dane_enable(&mut self, base_domain: &str) -> Result<(), ErrorStack> {
+        let cstr = CString::new(base_domain).unwrap();
+        unsafe { cvt(ffi::SSL_dane_enable(self.as_ptr(), cstr.as_ptr())).map(|_| ()) }
+    }
+
+    /// May be called one or more times to load each of the TLSA records that apply
+    /// to the remote TLS peer.
+    ///
+    /// This must be done prior to the beginning of the SSL handshake.
+    ///
+    /// The arguments specify the fields of the TLSA record.
+    /// The data field is provided in binary (wire RDATA) form,
+    /// not the hexadecimal ASCII presentation form.
+    ///
+    /// Returns a boolean to indicate whether the record was usable or not.
+    ///
+    /// Requires OpenSSL 1.1.0 or newer.
+    #[corresponds(SSL_dane_tlsa_add)]
+    #[cfg(ossl110)]
+    pub fn dane_tlsa_add(
+        &mut self,
+        usage: DaneUsage,
+        selector: DaneSelector,
+        mtype: DaneMatchType,
+        data: &[u8],
+    ) -> Result<bool, ErrorStack> {
+        let usable = unsafe {
+            cvt_n(ffi::SSL_dane_tlsa_add(
+                self.as_ptr(),
+                usage.as_raw(),
+                selector.as_raw(),
+                mtype.as_raw(),
+                data.as_ptr() as *const c_uchar,
+                data.len(),
+            ))
+        }?;
+
+        Ok(usable > 0)
```

Comment:
Is an error stack not returned in the 0 case? I think this might be better off just returning `Result<(), ErrorStack>`.